iocore/net/P_SSLSNI.h

```diff
@@ -0,0 +1,85 @@
+/** @file
+
+  A brief file description
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+/*************************** -*- Mod: C++ -*- ******************************
+  P_SSLSNI.h
+   Created On      : 05/02/2017
+
+   Description:
+   SNI based Configuration in ATS
+ ****************************************************************************/
+#ifndef __P_SSL_SNI_H__
+#define __P_SSL_SNI_H__
+
+#include "ProxyConfig.h"
+#include "ts/Map.h"
+#include "P_ActionProcessor.h"
+#include "ts/MatcherUtils.h"
+#include "openssl/ossl_typ.h"
+#include <vector>
+#include <strings.h>
+
+// Properties for the next hop server
+struct NextHopProperty {
+  const char *name   = nullptr; // name of the server
+  int8_t verifyLevel = 0;       // whether to verify the next hop
+  SSL_CTX *ctx       = nullptr; // ctx generated off the certificate to present to this server
+  NextHopProperty();
+};
+
+typedef std::vector<ActionItem *> actionVector;
+typedef HashMap<cchar *, StringHashFns, actionVector *> SNIMap;
+typedef HashMap<cchar *, StringHashFns, NextHopProperty *> NextHopPropertyTable;
+
+struct SNIConfigParams : public ConfigInfo {
+  char *sni_filename = nullptr;
+  mutable SNIMap sni_action_map;
```

Comment:
The reason this is mutable is because all the getter functions of HashMap or Map in lib/ts are non-const so I can't pass the keys (which are cchar*) to obtain the mapped objects.